### PR TITLE
Added layer norm to policy network and entropy bonus to policy loss.

### PR DESCRIPTION
**Pull request by dosssman**
Glad if I could have your input on the entropy computation and how it is added to the policy loss.

Sorry for the # DEBUG lines, but I have to run right now.

### CODE REVIEW

File: cleanrl/ppo2_continuous_action.py [hunk 1]
```diff
@@ -232,6 +258,31 @@ def get_logproba(self, x, actions):
...
+
+        # print( "# DEBUG Entropy shape")
+        # print( entropy.shape)
+
+        # print( "# DEBUG: Manual entropy computation")
+        # print( - logprobs)
+        # print( "# DEBUG: Pytorch entropy computation")
+        # print( dist.entropy().sum(1))
+
+        return entropies
```
Comment (vwxyzjn) [inline cleanrl/ppo2_continuous_action.py]: The `get_entropies` function could be combined with
```python 
    def get_action(self, x):
        mean, logstd = self.forward(x)
        std = torch.exp(logstd)
        probs = Normal(mean, std)
        action = probs.sample()
        return action, probs.log_prob(action).sum(1)

    def get_logproba(self, x, actions):
        # Note: Converts actions to tensor
        if not isinstance(actions, torch.Tensor):
            actions = preprocess_obs_fn(actions)

        action_mean, action_logstd = self.forward(x)
        action_std = torch.exp(action_logstd)
        dist = Normal(action_mean, action_std)
        return dist.log_prob(actions).sum(1)
```

and get something like this:

```python
    def get_action(self, x, action=None):
        mean, logstd = self.forward(x)
        std = torch.exp(logstd)
        probs = Normal(mean, std)
        action = probs.sample() if not action
        return action, probs.log_prob(action).sum(1), dist.entropy().sum(1)
```

If you decided to go with Manual entropy computation, then do the entropy like in https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_continuous_action.py
```python
        newlogproba = pg.get_logproba(obs[:step+1], torch.Tensor(actions[:step+1]))
        # newvalues = vf.forward(obs[:step+1]).flatten() DO we generate a new values from the current policy?
        ratio =  torch.exp(newlogproba - torch.Tensor(logprobs[:step+1]))
        surrogate1 = ratio * torch.Tensor(advantages[:step+1])
        surrogate2 = ratio.clamp(1 - args.clip_coef, 1 + args.clip_coef) * torch.Tensor(advantages[:step+1])
        policy_loss = - torch.mean(torch.min(surrogate1, surrogate2))
        vf_loss = torch.mean((values[:step+1] - torch.Tensor(returns[:step+1])).pow(2))

        entropy_loss = torch.mean(torch.exp(newlogproba) * newlogproba)

        total_loss = policy_loss + args.vf_coef * vf_loss + args.ent_coef * entropy_loss
        optimizer.zero_grad()
        total_loss.backward(retain_graph=True)
```

Also, I don't quite see why `rsample()` is necessary here.

File: cleanrl/ppo2_continuous_action.py [hunk 2]
```diff
@@ -378,7 +439,13 @@ def forward(self, x):
                                 (1.+args.clip_coef) * advantages,
                                 (1.-args.clip_coef) * advantages).to(device)
 
-        policy_loss = - torch.min(ratio * advantages, clip_adv).mean()
+        # Adding entropy bonus: This is computed using the logprobs and the rsampled actions
+        # so the gradients can also be back propagated through it (?)
+        # Does the entropy function of a distribution backpropagate ? Probably not.
```
Comment (vwxyzjn) [inline cleanrl/ppo2_continuous_action.py]: Yes. See 
image

File: cleanrl/ppo2_continuous_action.py [hunk 3]
```diff
@@ -378,7 +439,13 @@ def forward(self, x):
...
+        # Also, this method does not used the old action sampled from and old policy to get the log probs
+        # Any discussion on what would be the difference ?
```
Comment (vwxyzjn) [inline cleanrl/ppo2_continuous_action.py]: If you are using Pytorch entropy computation, then the sampled actions has nothing to do with the entropy, which itself is calculated solely by the standard deviation of the normal distribution.

File: cleanrl/ppo2_continuous_action.py (hunk 1 above)
Comment (dosssman) [reply cleanrl/ppo2_continuous_action.py]: Thanks for the suggestions.

I have run some experiments with both methods for the entropy computation and while the average results are mostly similar, the manual computed entropy seems better, as per the following report : [WANDB Report PPO2 Entropy Computation](https://www.reddit.com/r/MachineLearning/comments/f5kkxb/discussion_lessons_learned_from_my_failures_in/?utm_medium=android_app&utm_source=share).

Also moved the entropy computation to the network update section.

File: cleanrl/ppo2_continuous_action.py (hunk 3 above)
Comment (dosssman) [reply cleanrl/ppo2_continuous_action.py]: Thanks. It makes more sense now.
Anyway, let's use the manual computation, as OpenAI Baseline do the same.

File: cleanrl/ppo2_continuous_action.py (hunk 2 above)
Comment (dosssman) [reply cleanrl/ppo2_continuous_action.py]: Thanks. I am so lazy, as always hahaha.
Now I know it is possible to easily instanciate a `Variable` to create distributions and such.

File: cleanrl/ppo2_continuous_action.py [hunk 4]
```diff
@@ -208,8 +216,19 @@ def __init__(self):
...
+            x = self.ln1(x)
+        x = torch.tanh(x)
+
+        x = self.fc2(x)
+
+        if args.pol_layer_norm:
+            x = self.ln2(x)
+
+        x = torch.tanh(x)
+
```
Comment (vwxyzjn) [inline cleanrl/ppo2_continuous_action.py]: Delete the spaces, please. They seem fairly random.

File: cleanrl/ppo2_continuous_action.py [hunk 5]
```diff
@@ -300,6 +319,10 @@ def forward(self, x):
         values[step] = vf.forward(obs[step:step+1])
 
         with torch.no_grad():
+            # DEBUG: Tracking the logits
+            logits = pg(obs[step:step+1])
+            logits = logits[0][0].cpu().numpy()
+
```
Comment (vwxyzjn) [inline cleanrl/ppo2_continuous_action.py]: This shouldn't be here as this is in the `cleanrl` official folder. I do think they should be added to a separate file like in `cleanrl/experiments/ppo2_continuous_action_debug.py`

File: cleanrl/ppo2_continuous_action.py (hunk 4 above)
Comment (dosssman) [reply cleanrl/ppo2_continuous_action.py]: My bad. I was probably "high" when working on that part hahaha
